### ISSUE TITLE
How to get number of writable bytes in send buffer

### ISSUE THREAD

**Issue by Josh798** [LiamBindle/MQTT-C#35]
Hi Liam,

Is there a way to get the number of writable bytes remaining in a client's send buffer? I see that there is the [`curr_sz`](https://github.com/LiamBindle/MQTT-C/blob/f4fcafeb0398572d6aa8613b76f4fe21f0de87ee/include/mqtt.h#L969) field in the client's message queue, but this doesn't appear to do exactly what I need it to. This field will remain unchanged (as far as I can tell) after a call to `mqtt_sync`. 

Say my buffer size is 4096. Initially, `curr_sz` will be approximately 4096 (let's say 4000). After I publish a 150-byte message, `curr_sz` will decrease by just over 150 bytes. So now, `curr_sz` is somewhere around 3800. After I call `mqtt_sync`, `curr_sz` is still 3800. What I'm looking for is a field (or an expression, perhaps) that will give me 4000 after calling `mqtt_sync`, indicating that there is more room in the buffer because a message has been sent.

Basically, I'm trying to avoid the MQTT_ERROR_SEND_BUFFER_IS_FULL error. I am using the automatic recovery feature. But for my use case, it would be better for my application to not attempt to send a message that won't fit in the buffer.

Thanks for your help. Fantastic library, by the way.

**Comment by LiamBindle** [comment]
Hi Josh,

Yeah, you are correct that `curr_sz` in `struct mqtt_client`  doesn't usually change when `mqtt_sync` is called. That is because the buffer is only flushed when you try to add a packet that is too big to currently fit. This prevents the need to flush of acknowledged packets after each received acknowledgement (a flush requires a memcpy of all unacknowledged packets in the buffer).

I think that `curr_sz` in `struct mqtt_message_queue` is what you're looking for ([here](https://liambindle.ca/MQTT-C/structmqtt__message__queue.html)). Does that look right to you?

Cheers,

Liam

**Comment by Josh798** [comment]
Yes, the `curr_sz` field in `struct mqtt_message_queue` is what I was referring to. At a first glance, that appears to be what I need. But when I actually use it, it doesn't change after calling `mqtt_sync`. I created a quick & dirty paste [here](https://pastebin.com/UaYJnyAS) that should demonstrate this behavior. When I compile and run this code, this is the output I see:
```
Initially, curr_sz = 1942
After publishing, curr_sz = 1870
After syncing, curr_sz = 1870
```

**Comment by LiamBindle** [comment]
Oh yeah, I see what you mean. Unfortunately, I don't think there is a way to getting what you are looking for without looping through the start of the queue to see how many messages are of state `MQTT_QUEUED_COMPLETE`.

Try calling `mqtt_mq_clean(client->mq)` after every `mqtt_sync(client)`. This should force the message queue to be cleaned immediately, and bring the `curr_sz` field to the maximum possible number of bytes that can be sent.

Let me know if you still have any problems.

**Comment by Josh798** [comment]
Okay, yeah that seems like it will do the trick. I've got a busy couple weeks ahead of me so I haven't actually tried it yet, but it looks promising. I'll close this for now and add a comment or re-open when I actually try it out. 

Thanks for your help,
Josh

**Comment by LiamBindle** [comment]
Great, sounds good. Let me know how it goes when you get to it.

Liam

**Comment by Josh798** [comment]
I finally got around to this this week. This generally did what I wanted it to. 

An issue I ran into, however, was that `__mqtt_ping` could be called on a call to `mqtt_sync`, which could cause an `MQTT_ERROR_SEND_BUFFER_IS_FULL` error. This is the error I was trying to avoid in the first place. This scenario would _probably_ only happen if a broker silently disappears when the send buffer is almost completely full of unacknowledged data. I'm working on putting together a system that needs to be relatively fault-tolerant, so my workaround was to make sure that before I call `mqtt_publish` (or any other function that puts data on the send buffer), that `client->mq.curr_sz` is larger than:

1. The size of the packet I'm trying to put on the buffer, PLUS
2. The size of the ping packet which _could_ be placed on the buffer, PLUS
3. The size of a `mqtt_queued_message` struct used for my packet, PLUS
4. The size of a `mqtt_queued_message` struct which _could_ be used for the ping packet.

Basically, I just make sure I always leave room for a possible ping packet before putting anything on the send buffer. Note that I am manually calculating points 1 & 2 above according to the MQTT 3.1.1 spec. I haven't seen the send-buffer-is-full error come across since doing what I've described here, so I think I've solved this issue.

Thanks!
Josh

**Comment by LiamBindle** [comment]
The safest thing is to make your buffers as big as you can afford, and use MQTT-C's reconnect functionality.  

Because pings, acks, etc are all happening behind the scenes, and being put on the same send buffer that your publishes are put on, checking that  `client->mq.curr_sz` is smaller than your publish message doesn't mean the buffer is big enough.

MQTT-C really does try to be semantically correct, so I'd caution against inspecting the maximum usage of your send buffer for anything besides finding a "reasonable" buffer size for your application.
